project_task_materials_stock/models/project.py

```diff
@@ -0,0 +1,169 @@
+# -*- coding: utf-8 -*-
+# (c) 2015 Antiun Ingeniería S.L. - Sergio Teruel
+# (c) 2015 Antiun Ingeniería S.L. - Carlos Dauden
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+from openerp import models, fields, api
+from openerp.tools.float_utils import float_round
+
+
+class ProjectTaskType(models.Model):
+    _inherit = 'project.task.type'
+
+    consume_material = fields.Boolean(
+        string='Consume Material',
+        help="If you mark this check, when a task goes to this state,"
+             "it will consume the associated materials")
+
+
+class Task(models.Model):
+    _inherit = "project.task"
+
+    def _compute_stock_move(self):
+        move_ids = [line.stock_move_id.id for line in self.material_ids
+                    if line.stock_move_id]
+        self.stock_move_ids = self.env['stock.move'].browse(move_ids)
+
+    def _compute_analytic_line(self):
+        line_ids = [line.analytic_line_id.id for line in self.material_ids
+                    if line.analytic_line_id]
+        self.analytic_line_ids = \
+            self.env['account.analytic.line'].browse(line_ids)
+
+    @api.depends('stock_move_ids.state')
+    def _check_stock_state(self):
+        if not self.stock_move_ids:
+            self.stock_state = 'pending'
+        elif self.stock_move_ids.filtered(lambda r: r.state == 'confirmed'):
+            self.stock_state = 'confirmed'
+        elif self.stock_move_ids.filtered(lambda r: r.state == 'assigned'):
+            self.stock_state = 'assigned'
+        elif self.stock_move_ids.filtered(lambda r: r.state == 'done'):
+            self.stock_state = 'done'
+
+    stock_move_ids = fields.One2many(
+        comodel_name='stock.move', compute='_compute_stock_move',
+        string='Stock Moves')
+    analytic_line_ids = fields.One2many(
+        comodel_name='account.analytic.line', compute='_compute_analytic_line',
+        string='Analytic Lines')
+    consume_material = fields.Boolean(related='stage_id.consume_material')
+    stock_state = fields.Selection(
+        [('pending', 'Pending'),
+         ('confirmed', 'Confirmed'),
+         ('assigned', 'Assigned'),
+         ('done', 'Done')], compute='_check_stock_state', string='Stock State')
+
+    @api.one
+    def unlink_stock_move(self):
+        for move in self.stock_move_ids:
+            if move.state == 'assigned':
+                move.do_unreserve()
+            if move.state in ['waiting', 'confirmed', 'assigned']:
+                move.state = 'draft'
+            move.unlink()
+
+    @api.one
+    def write(self, vals):
+        res = super(Task, self).write(vals)
+        if 'stage_id' in vals:
+            if self.consume_material:
```

Comment:
The aim of this is that the moves are only created or deleted when the stage of the task has been modified.
Otherwise moves would be created in each write.